src/passes/Unsubtyping.cpp

```diff
@@ -0,0 +1,580 @@
+          if (HeapType::isSubType(type, dest)) {
+            noteSubtype(type, dest);
+          }
+        }
+      }
+    }
+  }
+
+  void optimizeTypes(Module& wasm) {
+    struct Rewriter : GlobalTypeRewriter {
+      Unsubtyping& parent;
+      Rewriter(Unsubtyping& parent, Module& wasm)
+        : GlobalTypeRewriter(wasm), parent(parent) {}
+      std::optional<HeapType> getSuperType(HeapType type) override {
+        if (auto it = parent.supertypes.find(type);
+            it != parent.supertypes.end() && !it->second.isBasic()) {
+          return it->second;
+        }
+        return std::nullopt;
+      }
+    };
+    Rewriter(*this, wasm).update();
+  }
+
+  void doWalkModule(Module* wasm) {
+    // Visit the functions in parallel.
+    ModuleUtils::ParallelFunctionAnalysis<Unsubtyping> analysis(
+      *wasm, [&](Function* func, Unsubtyping& unsubtyping) {
+        if (!func->imported()) {
+          unsubtyping.walkFunctionInModule(func, wasm);
+        }
+      });
+    // Collect the results from the functions.
+    for (auto& [_, unsubtyping] : analysis.map) {
+      for (auto [sub, super] : unsubtyping.supertypes) {
+        noteSubtype(sub, super);
+      }
+      for (auto& [src, dests] : unsubtyping.castTypes) {
+        for (auto dest : dests) {
+          noteCast(src, dest);
+        }
+      }
+    }
+    // Collect constraints from top-level items.
+    for (auto& global : wasm->globals) {
+      visitGlobal(global.get());
+    }
+    for (auto& seg : wasm->elementSegments) {
+      visitElementSegment(seg.get());
+    }
+    // Visit the rest of the code that is not in functions.
+    walkModuleCode(wasm);
+  }
+
+  void visitFunction(Function* func) {
+    if (func->body) {
+      noteSubtype(func->body->type, func->getResults());
+    }
+  }
+  void visitGlobal(Global* global) {
+    if (global->init) {
+      noteSubtype(global->init->type, global->type);
+    }
+  }
+  void visitElementSegment(ElementSegment* seg) {
+    if (seg->offset) {
+      noteSubtype(seg->type, getModule()->getTable(seg->table)->type);
+    }
+    for (auto init : seg->data) {
+      noteSubtype(init->type, seg->type);
+    }
+  }
+  void visitNop(Nop* curr) {}
+  void visitBlock(Block* curr) {
+    if (!curr->list.empty()) {
+      noteSubtype(curr->list.back()->type, curr->type);
+    }
+  }
+  void visitIf(If* curr) {
+    if (curr->ifFalse) {
+      noteSubtype(curr->ifTrue->type, curr->type);
+      noteSubtype(curr->ifFalse->type, curr->type);
+    }
+  }
+  void visitLoop(Loop* curr) { noteSubtype(curr->body->type, curr->type); }
+  void visitBreak(Break* curr) {
+    if (curr->value) {
+      noteSubtype(curr->value->type, findBreakTarget(curr->name)->type);
+    }
+  }
+  void visitSwitch(Switch* curr) {
+    if (curr->value) {
+      noteSubtype(curr->value->type, findBreakTarget(curr->default_)->type);
+      for (auto name : curr->targets) {
+        noteSubtype(curr->value->type, findBreakTarget(name)->type);
```

Comment:
Better to use `BranchUtils::getUniqueTargets`, as (1) it's shorter, and (2) it deduplicates the names first, leading to less work (often many switch targets are identical).